server/src-lib/Hasura/Eventing/EventTrigger.hs

```diff
@@ -100,94 +76,116 @@ defaultFetchIntervalMilliSec :: (Milliseconds 'Absolute)
-          timeoutSeconds = fromMaybe defaultTimeoutSeconds (rcTimeoutSec retryConf)
-          respTimeout = HTTP.responseTimeoutMicro (timeoutSeconds * 1000000)
-          headerInfos = etiHeaders eti
-          etHeaders = map encodeHeader headerInfos
-          headers = addDefaultHeaders etHeaders
-          ep = createEventPayload retryConf e
-          extraLogCtx = ExtraLogContext (epId ep)
-      res <- runExceptT $ withEventEngineCtx eventEngineCtx $
-        tryWebhook headers respTimeout (toJSON ep) webhook
-      logHTTPForET res (Just extraLogCtx)
-      let decodedHeaders = map (decodeHeader logenv headerInfos) headers
-      finally <- either
-        (processError pool e retryConf decodedHeaders ep)
-        (processSuccess pool e decodedHeaders ep) res
-      either logQErr return finally
-
-withEventEngineCtx ::
-  ( MonadIO m
-  , MonadMask m
-  )
-  => EventEngineCtx
-  -> m (HTTPResp a)
-  -> m (HTTPResp a)
-withEventEngineCtx eeCtx = bracket_ (incrementThreadCount eeCtx) (decrementThreadCount eeCtx)
-
-incrementThreadCount :: MonadIO m => EventEngineCtx -> m ()
-incrementThreadCount (EventEngineCtx _ c maxT _ ) = liftIO $ atomically $ do
-  countThreads <- readTVar c
-  if countThreads >= maxT
-    then retry
-    else modifyTVar' c (+1)
-
-decrementThreadCount :: MonadIO m => EventEngineCtx -> m ()
-decrementThreadCount (EventEngineCtx _ c _ _) = liftIO $ atomically $ modifyTVar' c (\v -> v - 1)
+    fetchBatchSize = 100
+    popEventsBatch = do
+      let run = runExceptT . Q.runTx pool (Q.RepeatableRead, Just Q.ReadWrite)
+      run (fetchEvents fetchBatchSize) >>= \case
+          Left err -> do
+            L.unLogger logger $ EventInternalErr err
+            return []
+          Right events ->
+            return events
+
+    -- work on this batch of events while prefetching the next. Recurse after we've forked workers
+    -- for each in the batch, minding the requested pool size.
+    go :: [Event] -> Int -> Bool -> IO void
+    go events !fullFetchCount !alreadyWarned = do
+      -- process events ASAP until we've caught up; only then can we sleep
+      when (null events) $ sleep _eeCtxFetchInterval
+
+      -- Prefetch next events payload while concurrently working through our current batch.
+      -- NOTE: we probably don't need to prefetch so early, but probably not
+      -- worth the effort for something more fine-tuned
+      eventsNext <- withAsync popEventsBatch $ \eventsNextA -> do
+        -- process approximately in order, minding HASURA_GRAPHQL_EVENTS_HTTP_POOL_SIZE:
+        forM_ events $ \event ->
+          mask_ $ do
+            atomically $ do  -- block until < HASURA_GRAPHQL_EVENTS_HTTP_POOL_SIZE threads:
+              capacity <- readTVar _eeCtxEventThreadsCapacity
+              check $ capacity > 0
+              writeTVar _eeCtxEventThreadsCapacity (capacity - 1)
+            -- since there is some capacity in our worker threads, we can launch another:
```

Comment:
It would be great to use the `bracket` pattern here for incrementing and decrementing counters